src/2d/shallow/landspill/point_source/point_source_module.f90

```diff
@@ -0,0 +1,296 @@
+    end function constructor_direct
+
+    !> @brief A mimic to C++ stype constructor.
+    !! @param[in] funit the unit of the file.
+    !! @return a new instance with provided setup.
+    function constructor_funit(funit)
+        ! variable declaration
+        type(PointSource):: constructor_funit
+        integer(kind=4), intent(in):: funit
+
+        call init_from_funit(constructor_funit, funit)
+    end function constructor_funit
+
+    !> @brief Finalize a PointSource instance.
+    !!
+    !! Deallocate dynamic arrays and finalize the instance.
+    !!
+    !! @param[in, out] this an instance of PointSource
+    subroutine destructor(this)
+        ! declaration
+        type(PointSource), intent(inout):: this
+
+        ! code
+        if (allocated(this%t)) deallocate(this%t)
+        if (allocated(this%v_rate)) deallocate(this%v_rate)
+
+        this%coord = (/ 0D0, 0D0 /)
+        this%nt = -1
+        this%id = -1
+
+    end subroutine destructor
+
+    !> @brief Initialize a PointSource instance.
+    !!
+    !! Initializ a PointSource instance.
+    !!
+    !! @param[in, out] this an instance of PointSource
+    !! @param[in] id ID/tag of this object.
+    !! @param[in] coord a 1D real(kind=8) array with size 2, representing x and y coordinates
+    !! @param[in] nt number of time segments
+    !! @param[in] t a 1D array of reals with size nt; times at the end of segments
+    !! @param[in] v_rate a 1D array of reals with size nt; values
+    subroutine init_direct(this, id, coord, nt, t, v_rate)
+        ! variable declaration
+        class(PointSource), intent(inout):: this
+        integer(kind=4), intent(in):: id
+        real(kind=8), dimension(2), intent(in):: coord
+        integer(kind=4), intent(in):: nt
+        real(kind=8), dimension(nt), intent(in):: t
+        real(kind=8), dimension(nt), intent(in):: v_rate
+
+        ! code
+        call destructor(this)
+
+        this%id = id
+        this%coord = coord
+        this%nt = nt
+        this%t = t ! implicit allocation
+        this%v_rate = v_rate ! implicit allocation
+
+        ! TODO: make sure the t array is sorted and v_rate is sorted based on t
+
+    end subroutine init_direct
+
+    !> @brief Initialize a PointSource instance with a file unit provided.
+    !!
+    !! Initializ a PointSource instance.
+    !!
+    !! @param[in, out] this an instance of PointSource
+    !! @param[in] funit unit of the file.
+    subroutine init_from_funit(this, funit)
+        ! variable declaration
+        class(PointSource), intent(inout):: this
+        integer(kind=4), intent(in):: funit
+
+        ! code
+        call destructor(this)
+
+        read(funit, *) this%id
+        read(funit, *) this%coord
+        read(funit, *) this%nt
+
+        allocate(this%t(this%nt), this%v_rate(this%nt))
+
+        read(funit, *) this%t
+        read(funit, *) this%v_rate
+
+        ! TODO: make sure the t array is sorted and v_rate is sorted based on t
```

Comment:
Just an FYI, there's a quicksort that is implemented already in AMRClaw that might be useful for sorting these.

Reply:
Yes, I think I found it. If I'm correct, that should be the subroutine `qsorti` in `quick_sort1.f`. I'll work on this.

Reply:
On second thought, I think if `t` array is not sorted, then it is an error of input data. So I believe checking whether a user-provided list is sorted or not should be done during the preprocessing stage. That is, the checking should happen when users do `make .data`.

So I will not do either sorting or checking here. I will do the checking in `data.py`, instead.

Reply:
Done in 61455d9.